### PR TITLE
Position the frames with DrawQuad functions

### PR DESCRIPTION
**Pull request by Namaneo**
Required by the upcoming zoom feature, which require contextual positioning of each quad.

`MTY_Position` has been created to define how a frame is displayed on the screen. It defaults to `MTY_POSITION_AUTO`, which is the original behavior. When `MTY_POSITION_FIXED` is set, `imageX` and `imageY` can be set to position the frame. In this mode, automatic scaling is disabled and must be set in the `scale` property by the client.

Note: In OpenGL, `windowWidth` and `windowHeight` must be set on `MTY_DrawDesc`. This is done automatically in `MTY_WindowDrawQuad`.

### CODE REVIEW

File: src/gfx/viewport.h [hunk 1]
```diff
@@ -8,36 +8,58 @@
 
 #include <math.h>
 
-static void mty_viewport(MTY_Rotation rotation, uint32_t w, uint32_t h,
-	uint32_t view_w, uint32_t view_h, float ar, float scale, float *vp_x,
-	float *vp_y, float *vp_w, float *vp_h)
+static void mty_viewport(const MTY_RenderDesc *desc, float *vp_x, float *vp_y, float *vp_w, float *vp_h, bool transform_origin)
 {
```
Comment (RonaldH) [inline src/gfx/viewport.h]: In general we don't want to pass around the struct

**Review by RonaldH — CHANGES_REQUESTED** [review]

File: src/gfx/viewport.h (hunk 1 above)
Comment (Namaneo) [reply src/gfx/viewport.h]: Agreed together to keep the new prototype

File: src/gfx/viewport.h [hunk 2]
```diff
@@ -8,36 +8,58 @@
...
+	float scaled_w = roundf(desc->scale * w);
+	float scaled_h = roundf(desc->scale * h);
 
-	if (scaled_w == 0 || scaled_h == 0 || view_w < scaled_w || view_h < scaled_h)
-		scaled_w = view_w;
+	switch (desc->type)	{
+		default:
+		case MTY_POSITION_AUTO:
+			if (scaled_w == 0 || scaled_h == 0 || view_w < scaled_w || view_h < scaled_h)
+				scaled_w = view_w;
```
Comment (RonaldH) [inline src/gfx/viewport.h]:
```suggestion
			if (scaled_w == 0.0f || scaled_h == 0.0f || view_w < scaled_w || view_h < scaled_h)
				scaled_w = view_w;
```

File: src/gfx/viewport.h [hunk 3]
```diff
@@ -8,36 +8,58 @@
...
 	}
 
-	uint32_t scaled_w = lrint(scale * (float) w);
-	uint32_t scaled_h = lrint(scale * (float) h);
+	float scaled_w = roundf(desc->scale * w);
+	float scaled_h = roundf(desc->scale * h);
 
-	if (scaled_w == 0 || scaled_h == 0 || view_w < scaled_w || view_h < scaled_h)
-		scaled_w = view_w;
+	switch (desc->type)	{
```
Comment (RonaldH) [inline src/gfx/viewport.h]:
```suggestion
	switch (desc->type) {
```

File: src/matoya.h [hunk 4]
```diff
@@ -99,18 +99,30 @@ typedef enum {
...
+	MTY_POSITION_MAKE_32 = INT32_MAX,
+} MTY_Position;
+
 /// @brief Description of a render operation.
 typedef struct {
 	MTY_ColorFormat format; ///< The color format of a raw image.
 	MTY_Rotation rotation;  ///< Rotation applied to the image.
 	MTY_Filter filter;      ///< Filter applied to the image.
 	MTY_Effect effect;      ///< Effect applied to the image.
+	MTY_Position type;      ///< Type of positioning configuration.
```
Comment (RonaldH) [inline src/matoya.h]:
```suggestion
	MTY_Position position;  ///< Type of positioning configuration.
```

**Review by RonaldH — CHANGES_REQUESTED** [review]

**Review by RonaldH — APPROVED** [review]

**Review by RonaldH — APPROVED** [review]

**Review by Kodikuu — APPROVED** [review]

**Pull request closed without merging**